### PR TITLE
Jimmy fixing sampled initial values for optimizers issue#1364

### LINKED ISSUE #1364: [DEFECT] Optimizers without initial values errors out if all variables are initialized through sampler.
--------
If all initial values of optimized variables are to be provided through a sampler an error occurs:
File "/Users/abdomg/projects/raven/framework/Optimizers/Optimizer.py", line 403, in _initializeInitSampler
    `numTraj = len(self._initialValues)`
TypeError: object of type 'NoneType' has no len()
--------

**Describe the defect**
Although there are two ways to provide initial values of an optimizer: user-provided values or sampled through a user-defined sampler, the latter does not work properly unless at least one variable has user-defined initial values.  

##### What did you expect to see happen?
no errors should occur

##### What did you see instead?
Raven errored out

##### Do you have a suggested fix for the development team?
Yes, 
replace line 403 in Optimizer.py from : `numTraj = len(self._initialValues)`
to: `numTraj = len(self._initialValues) if self._initialValues else None`
**Describe how to Reproduce**
1. just run the attached test.
**Screenshots and Input Files**
Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

**Platform (please complete the following information):**
 - Version: [e.g. 22]
 - Dependencies Installation: [CONDA or PIP]

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users' group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with the release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?
[init_sampler.xml.zip](https://github.com/idaholab/raven/files/5478594/init_sampler.xml.zip)

### PR DESCRIPTION
**Pull request by Jimmy-INL**
--------
This PR enables the initialization of all variables through a sampler in optimizers. It closes #1364
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
closes #1364 

##### What are the significant changes in functionality due to this change request?
Right now at least one of the optimized variables has to be initialized by the user. This trivial fix should 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: tests/framework/Optimizers/GradientDescent/init_sampler_allVars.xml [hunk 1]
```diff
@@ -0,0 +1,126 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>framework/Optimizers.ConvergeGradient</name>
```
Comment (alfoa) [inline tests/framework/Optimizers/GradientDescent/init_sampler_allVars.xml:4]: the name is not correct

it is 
framework/Optimizers/GradientDescent.InitSampler

File: tests/framework/Optimizers/GradientDescent/tests [hunk 2]
```diff
@@ -91,6 +91,18 @@
     [../]
   [../]
 
+  [./InitSampler]
+    type = 'RavenFramework'
+    input = 'init_sampler_allVars.xml'
+    [./data]
+      type = OrderedCSV
+      output = 'InitSamplerAllVars/opt_export_0.csv InitSamplerAllVars/opt_export_1.csv'
+
```
Comment (alfoa) [inline tests/framework/Optimizers/GradientDescent/tests:100]: remove empty line

**Review by alfoa — CHANGES_REQUESTED** [review]

File: tests/framework/Optimizers/GradientDescent/init_sampler_allVars.xml (hunk 1 above)
Comment (Jimmy-INL) [reply tests/framework/Optimizers/GradientDescent/init_sampler_allVars.xml:4]: Is that the case with the init_sampler.xml as well? I have copied it from there.

**Comment by alfoa** [comment]
@Jimmy-INL @PaulTalbot-INL , Mohammad, HERON is failing. do you know why?

**Comment by PaulTalbot-INL** [comment]
if it's Windows, we don't fully support windows on the testing environment yet. We have windows users running stuff, but the test machine fails on it still. Haven't had a spare 10 minutes to go find out why.

**Review by alfoa — APPROVED** [review]

**Comment by alfoa** [comment]
Checklist passed...

PR approved...

Merging....

**Pull request merged**